src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/xattention.hpp

```diff
@@ -0,0 +1,434 @@
+// Copyright (C) 2018-2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+#pragma once
+
+#include <fstream>
+#include <utility>
+
+#if defined(OPENVINO_ARCH_X86_64)
+#    include "nodes/kernels/x64/brgemm_kernel.hpp"
+#elif defined(OPENVINO_ARCH_ARM64)
+#    include "nodes/kernels/aarch64/brgemm_kernel.hpp"
+#endif
+#include "openvino/core/parallel.hpp"
+#include "openvino/core/shape.hpp"
+#include "openvino/core/type/element_type.hpp"
+#include "utils/plain_tensor.hpp"
+
+#if defined(HAVE_AVX2) || defined(HAVE_AVX512F)
+#    include <immintrin.h>
+
+#    include "common.hpp"
+#endif
+#include "softmax_kernel.hpp"
+#include "transpose_kernel.hpp"
+
+using namespace ov::intel_cpu;
+namespace ov::Extensions::Cpu::XARCH {
+
+template <typename TDST, typename TSRC>
+inline void transpose_tailx16_kernel(TDST* dst,
+                                     TSRC* src,
+                                     size_t n_cnt,
+                                     size_t k_cnt,
+                                     size_t dst_stride,
+                                     size_t src_stride) {
+    for (size_t i = 0; i < n_cnt; i++) {
+        for (size_t j = 0; j < k_cnt; j++) {
+            dst[j * dst_stride + i] = static_cast<TDST>(src[j + i * src_stride]);
+        }
+    }
+}
+
+template <typename TDST,
+          ov::element::Type_t SRC_PREC,
+          std::enable_if_t<(none_of(SRC_PREC, ov::element::i8, ov::element::u8, ov::element::u4)), bool> = true>
+void transpose_16NxK(TDST* dst,
+                     void* src,
+                     const size_t N,
+                     const size_t K,
+                     const size_t block_size,
+                     const size_t dst_stride,
+                     const size_t src_stride) {
```

Comment:
We have pretty similar templates in `src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp`. Can't we extract them in a common header `transpose.hpp` and reuse across implementations?

Reply:
At first glance we should keep a generic implementation with tail processing. Simple condition check for tail shouldn't impose any noticeable performance overhead. I don't mind if we handle it in a separate PR, but please create this follow-up PR right now so we don't lose this refactoring.

Reply:
Moved transpose functions to transpose.hpp. Will refactor executor_pa.cpp to use this header once the PR is merged. Created a ticket to track this: https://jira.devtools.intel.com/browse/CVS-177312.